### PR TITLE
Handled/unhandled error state serialisation fix

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal
Handled error data serialisation is synchronised through the use of `NSLock`. However, this does not apply to unhandled errors, which suspend all threads and can therefore unintentionally serialise data belonging to a handled error. 

This fix alters serialisation of error data by ignoring some unnecessary fields for unhandled errors, and calculating `severity` and other items using existing data.

### Added
- New internal typedef for a report write callback, which passes information on the whether a report is handled/unhandled.

### Changed
Ignore serialised values on unhandled errors for:

- depth
- severity
- handledStateJSON
- metaDataJSON
- userOverridesJSON

Calculate severity/depth using existing information for unhandled errors in `BugsnagCrashReport`

## Tests

New unit tests for serialisation of report depth + severity, mazerunner run on CI, manual testing to ensure that one handled/unhandled error were received.

### Outstanding Questions

- We may want to make `BSGReportCallback` public and deprecated `onCrashHandler`. I'm not 100% convinced that the information available in context makes this worth doing as it can mostly be obtained by other means, so I've left as-is for now.
- I'm fairly sure that depth, severity, and handledStateJSON don't need to be serialised, but it would be good to confirm that this doesn't cause some unintended side effect or misses an edge-case.

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: Source/BugsnagNotifier.m [hunk 1]
```diff
@@ -81,40 +82,41 @@
...
         writer->addJSONElement(writer, "state", bsg_g_bugsnag_data.stateJSON);
     }
-    if (bsg_g_bugsnag_data.userOverridesJSON) {
-        writer->addJSONElement(writer, "overrides",
-                               bsg_g_bugsnag_data.userOverridesJSON);
+
+    // write additional user-supplied metadata
+    if (userReported) {
+        if (bsg_g_bugsnag_data.metaDataJSON) {
+            writer->addJSONElement(writer, "metaData", bsg_g_bugsnag_data.metaDataJSON);
```
Comment (kattrali) [inline Source/BugsnagNotifier.m]: This value should be attached to all reports as it contains the values from [`+[Bugsnag addAttribute:withValue:toTabWithName:]`](https://docs.bugsnag.com/platforms/ios/customizing-error-reports/#adding-custom-diagnostics). We should add a test for this case.

**Comment by fractalwrench** [comment]
I've updated this so that the metadata JSON is always included in reports. What would be the best way to test this? Mazerunner for Cocoa is currently out of action on my local machine.

File: Tests/BugsnagCrashReportTests.m [hunk 2]
```diff
@@ -449,4 +449,35 @@ - (void)testEnhancedErrorMessageIgnoresUnknownAssertionTypes {
...
+    XCTAssertEqual(report.depth, 2);
+}
+
+- (void)testReportSeverity {
+    // unhandled reports should calculate their own severity
+    NSDictionary *dict = @{@"user.state.crash.severity": @"info"};
+    BugsnagCrashReport *report = [[BugsnagCrashReport alloc] initWithKSReport:dict];
+    XCTAssertEqual(report.severity, BSGSeverityError);
+
+    // handled reports should use the serialised depth
```
Comment (snmaynard) [inline Tests/BugsnagCrashReportTests.m]: This comment needs updating. I would also split these tests into 2 if it was me.

**Comment by snmaynard** [comment]
You could add a unit test for the metadata thing, similar to the severity tests you have. Whats wrong with mazerunner though?

**Comment by fractalwrench** [comment]
Added a unit test for metadata JSON serialisation as requested, and split some of the newly added tests into two.

Note that I can't seem to dismiss @kattrali's old review, so am unable to merge.
